### PR TITLE
Add extra naming inspections to the INFO_XML rules

### PR DESCRIPTION
**Pull request by dryabov**
### Summary of Changes
Add extra naming inspections to the INFO_XML rules

New checks:
- check manifest file do exist
- check [naming rules](https://extensions.joomla.org/support/knowledgebase/for-jed-developers/extension-names/)

Code changes:
- fixed loading of the language file
- drop Joomla!1.5 support ("install" root element)

### Testing Instructions
Install, upload any extension (e.g. anyone with Joomla in title) and click "Check" button

### Actual result BEFORE applying this Pull Request
No such a checks are available at all.

### Expected result AFTER applying this Pull Request
Naming issue(s) found.

### Documentation Changes Required
None (as there is no official documentation page with detailed info on every rules)

### CODE REVIEW

File: administrator/components/com_jedchecker/libraries/rules/xmlinfo.php
Comment (Llewellynvdm) [inline administrator/components/com_jedchecker/libraries/rules/xmlinfo.php:107]: We need more documentation here.

**Comment by Llewellynvdm** [comment]
Okay here are to many changes for me to just read and see if all is well. I will need to test this. Will take a little longer... for the most part it looks okay.

But I need to understand the motivation to the change of very line... and for that there is to little comments. I know am asking much... but I am thinking of those who are trying to understand the code, and does not know the Joomla API that well.

So even for testing better comments help other to see what you are trying to do, and make sure the outcome is as expected. Generally our testers are not advance programmers.

**Comment by conconnl** [comment]
The approved domains part is a bit strange for me, not the suggested change but the check in general. 

Approved domains are managed by the Trademark team and the list is changing everyday. There is even a registration component being build to replace the excel list and make it possible to show the list to the whole community. There is always a change that domains get removed. 

Maintaining a list on 2 places is unnecessary work and doing a extension coding check, which also holds a domain list check are too different worlds. 

Doing an informational check on the usage of the word joomla in the domain and telling the developer that they need to make sure they have Trademark approval is the only thing we could do in my opinion, unless the Checker uses live data from the trademark listing.

**Comment by dryabov** [comment]
@conconnl I prefer to avoid false-positives whenever it is possible, that's why that list is included. And it is not updated daily (e.g. there is just a couple of domains dated last year).
I've found a way to download that list as CSV

https://docs.google.com/spreadsheets/u/2/d/e/2PACX-1vTPafvM-uHf2qYvlfVkZ-pLl_dK6V12k9dpj0fqbXNqxWA-zArPoeHITIDZrirza9cxJWbh_2CAUmlS/pub?output=csv

so, it would be possible to download/update the list automatically, but I'm against the harcoding of this link in the code. 

Probably, the best solution would be to add a script on extensions.joomla.org that will download that file daily (by extracting link from https://tm.joomla.org/approved-domains.html to be sure it is actual), and put it to a public directory (or push to GitHub pages), so that JEDChecker will be able to download it.

If none suits you, it's not a problem to revert changes back.

**Comment by conconnl** [comment]
You exactly show the problem of the Excel file, it’s not complete, fully updated. 
Then again, I wouldn’t know why an extension checker would check trademark infringements. 

Sharing information with a developer is important of course. A check to see if on the (to be defined) places a trademarked name or domain is used and then sharing what to do (if not done yet) contacting Trademark is something without impact. 
@Llewellynvdm

**Comment by dryabov** [comment]
@conconnl Removed check against approved domain names. Anyway, it can be merged later, if necessary.

**Comment by Llewellynvdm** [comment]
Okay far as I know the domain list here https://tm.joomla.org/approved-domains.html is the official list of approved domains. What will be a mistake is to add a statics list to the extension, as this list can change anytime. The JED has the responsibility to not list extensions who are in breach of the trademark agreement with OSM. Therefore this addition could be helpful, but we will need to make sure of few things:

- That the list here https://tm.joomla.org/approved-domains.html is indeed complete, and if not, it must be made complete.
- That we link directly to google/drive/file via an API call to validate against the list to account for any changes.
- That the check will only give a warning if a problem is detected... and not an error. As the list may be behind... and we need to follow up the issue with @jeckodevelopment, or Claire.

Since I got [JoomlaComponentBuilder.com](http://joomlacomponentbuilder.com/) approved in 2018, and it only got added last week to the list... and so I can see that this issue may also need attention, but is part of the Trademark teams responsibility. So we need their input on this as well.

**Comment by conconnl** [comment]
@Llewellynvdm like I said, a new component will become available on the Trademark website which will hold the correct list. The excel file isn’t complete. 

Maybe @JazParkyn could add an API too it, as she is building the component to register all trademark related approvals, not just domains.

**Comment by dryabov** [comment]
> That we link directly to google/drive/file via an API call to validate against the list to account for any changes.

@Llewellynvdm The problem is that direct link to Google's Spreadsheet is not an official API. They can change URL of embedded stylesheet whenever they like. So, if we do rely on that list, we have to download https://tm.joomla.org/approved-domains.html page, extract URL of Google's Stylesheet document, convert it to URL for CSV format, download that file, parse it and save to a cache. It takes long time, and I wouldn't like to get XML_INFO rule result to be loaded the last. In this case it would be better to extract naming-related checks into a separate ruleset, keeping "info" ruleset to display just an info only (most likely it was the original intent of XML_INFO rule).

**Comment by Llewellynvdm** [comment]
This whole idea will have to wait until we have a better way to get the list of domains. So I am closing this PR, and will bump it again when the new list spoken by @conconnl is ready.

**Comment by Llewellynvdm** [comment]
Please move part that are not related to this issue to smaller PR that are easier to test by none technical volunteers.

**Comment by Llewellynvdm** [comment]
My apology... I see you did remove the list of domains... I will go over this code again.

File: administrator/components/com_jedchecker/libraries/rules/xmlinfo.php
Comment (Llewellynvdm) [inline administrator/components/com_jedchecker/libraries/rules/xmlinfo.php:112]: Can we not try to catch this to show the user that the xml is outdated?

File: administrator/components/com_jedchecker/libraries/rules/xmlinfo.php
Comment (Llewellynvdm) [inline administrator/components/com_jedchecker/libraries/rules/xmlinfo.php:150]: // Plugin's element name starts with plg_

File: administrator/components/com_jedchecker/libraries/rules/xmlinfo.php
Comment (Llewellynvdm) [inline administrator/components/com_jedchecker/libraries/rules/xmlinfo.php:234]: This regex only targets seems to narrow, it should instead look something like this which is more greedy.
```php
if (preg_match('/(module|plugin)/i', $extension_name))
```
Here is the idea https://regexr.com/5p86l

Let me know if this is to greedy ;)

Comment (dryabov) [reply administrator/components/com_jedchecker/libraries/rules/xmlinfo.php:234]: Initially I wasn't sure about other English words containing "module" or "plugin" as a substring, and `\b`s were added to avoid possible false-positives. But I've just checked at https://github.com/dwyl/english-words that it seems to be safe to check it "greedy". I'll commit this change a little bit later.

File: administrator/components/com_jedchecker/libraries/rules/xmlinfo.php
Comment (dryabov) [reply administrator/components/com_jedchecker/libraries/rules/xmlinfo.php:112]: What do you prefer (in addition to the warning message): to process `<install>` manifests anyway, or to don't? I prefer to don't process at all (and don't accept by JED) because `<extension>` is supported since J!1.6 and `<install>` is not supported by J!4.0 (I assume support of Joomla!4 will be required for new extensions shortly after J!4.0 release).

File: administrator/components/com_jedchecker/libraries/rules/xmlinfo.php
Comment (dryabov) [reply administrator/components/com_jedchecker/libraries/rules/xmlinfo.php]: Done

File: administrator/components/com_jedchecker/libraries/rules/xmlinfo.php
Comment (dryabov) [reply administrator/components/com_jedchecker/libraries/rules/xmlinfo.php]: Fixed

File: administrator/components/com_jedchecker/libraries/rules/xmlinfo.php
Comment (Llewellynvdm) [reply administrator/components/com_jedchecker/libraries/rules/xmlinfo.php:112]: Well currently if I follow this test correctly, if the `<extension>` is not found then the output will say xml not found. This is frustrating as the developer would say... #!&%@$% it's there... lol

So giving a more meaningful message that says the XML was found but is outdated implementation with a link to the new conventions.

But since it is happening during the search that we detect that the xml manifest convention is not matched. The next file being tested could be a match... so we can't show the message immediately. So we need to detect the `<install>` tag or any of the other older conventions and lineup a message... that only gets shown when the correct file is not found. So this message can say **XML not found since you are possibly not using the new convention in the manifest see URL for more details** or something like that.

Comment (Llewellynvdm) [reply administrator/components/com_jedchecker/libraries/rules/xmlinfo.php:112]: I need to check with @JazParkyn or @anibalsanchez where we are drawing the line... as about this I am not sure. But a more meaningful message will be a win. Not an error, but a warning for now... until I get some finality.

File: administrator/components/com_jedchecker/libraries/rules/xmlinfo.php
Comment (Llewellynvdm) [reply administrator/components/com_jedchecker/libraries/rules/xmlinfo.php]: That is a valid point... I will get back to you on this, maybe its to greedy now. I need to check the terms and make sure we don't make it to strict.

Comment (anibalsanchez) [reply administrator/components/com_jedchecker/libraries/rules/xmlinfo.php]: This is the list of banned words for extensions that are checked when the extension is submitted: 

- module
- plugin
- component
- free
- extension
- template

Any match of these words is rejected.

Also the word "Joomla" can't be in an extension name unless it is included at the end of the name " for Joomla"

Comment (anibalsanchez) [reply administrator/components/com_jedchecker/libraries/rules/xmlinfo.php]: One more, the extension name can't start with "mod_".

Comment (dryabov) [reply administrator/components/com_jedchecker/libraries/rules/xmlinfo.php]: I have an idea: let's trigger error for exact match (e.g. "Super Module"), warning for leading/trailing part of the word (e.g. "Supermodule"), and none (or a notice) for inner part of the word (e.g. "demoduled"). Maybe with exception of "free", because there are many words containing or ending with "free", and it's better to check for just an exact matching. What do you think?

@anibalsanchez Does JED allow the "package" word in the title? What about "plugin" written as "plug-in"? And what about other prefixes (com_, plg_, tpl_, pkg_)?

BTW, does JED allow emoji in extension name? (or "no longer" 😊?)

File: administrator/components/com_jedchecker/libraries/rules/xmlinfo.php
Comment (dryabov) [reply administrator/components/com_jedchecker/libraries/rules/xmlinfo.php:117]: A simple check has been committed

**Review by Llewellynvdm — APPROVED** [review]

**Pull request merged**
